src/Microsoft.Health.Fhir.Shared.Core/Features/Search/Preprocessors/ListSearchPreProcessor.cs

```diff
@@ -0,0 +1,104 @@
+        private const string _idParameter = "_id";
+        private readonly IScoped<IFhirDataStore> _dataStore;
+        private readonly ResourceDeserializer _deserializer;
+        private readonly IReferenceSearchValueParser _referenceSearchValueParser;
+
+        public ListSearchPreProcessor(
+            IScoped<IFhirDataStore> dataStore,
+            ResourceDeserializer deserializer,
+            IReferenceSearchValueParser referenceSearchValueParser)
+        {
+            EnsureArg.IsNotNull(dataStore, nameof(dataStore));
+            EnsureArg.IsNotNull(deserializer, nameof(deserializer));
+            EnsureArg.IsNotNull(referenceSearchValueParser, nameof(referenceSearchValueParser));
+
+            _dataStore = dataStore;
+            _deserializer = deserializer;
+            _referenceSearchValueParser = referenceSearchValueParser;
+        }
+
+        public async Task Process(SearchResourceRequest request, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(request, nameof(request));
+
+            Tuple<string, string> listParameter = request.Queries
+                .FirstOrDefault(x => string.Equals(x.Item1, _listParameter, StringComparison.Ordinal));
+
+            // Remove the 'list' params from the queries, to be later replaced with specific ids of resources
+            IEnumerable<Tuple<string, string>> query = request.Queries.Except(new[] { listParameter });
+
+            // if _list was not requested, or _list was requested with invalid value. continue
+            if ((listParameter == null) || string.IsNullOrWhiteSpace(listParameter.Item2))
+            {
+                request.Queries = query.ToArray();
+                return;
+            }
+
+            ResourceWrapper listWrapper = await _dataStore.Value.GetAsync(new ResourceKey<Hl7.Fhir.Model.List>(listParameter.Item2), cancellationToken);
+
+            if (listWrapper == null)
+            {
+                MakeAlwaysFalseQuery(ref query);
+                request.Queries = query.ToArray();
+                return;
+            }
+
+            ResourceElement list = _deserializer.Deserialize(listWrapper);
+
+            IEnumerable<ReferenceSearchValue> references = list.ToPoco<Hl7.Fhir.Model.List>()
+                .Entry
+                .Where(x => x.Deleted != true)
+                .Select(x => _referenceSearchValueParser.Parse(x.Item.Reference))
+                .Where(x => string.IsNullOrWhiteSpace(request.ResourceType) ||
+                            string.Equals(request.ResourceType, x.ResourceType, StringComparison.Ordinal))
+                .ToArray();
+
+            if (references.Any())
+            {
+                query = query.Concat(new[] { Tuple.Create(_idParameter, string.Join(",", references.Select(x => x.ResourceId))) });
+            }
+            else
+            {
+                MakeAlwaysFalseQuery(ref query);
+            }
+
+            request.Queries = query.ToArray();
```

Comment:
After reading more about the specifics of List, I think we also should consider how we can implement List order (https://www.hl7.org/fhir/list.html#order)

Reply:
Shouldn't that be in a different PR, one that touches the 'List' functionality (e.g. base/List) and not base/resource?_list=xyz ?
As I understand it the order in which is irrelevant (anyway I keep the order that the list had)